driver/src/snapp.rs

```diff
@@ -0,0 +1,409 @@
+    fn get_token_price(&self, token_id: u16) -> Option<u128> {
+        self.prices.get(token_id as usize).cloned()
+    }
+
+    fn snapp_objective_value(&self, orders: &[Order]) -> Result<U256, SnappObjectiveError> {
+        let mut total_executed_utility = U256::zero();
+        for (i, order) in orders.iter().enumerate() {
+            total_executed_utility = total_executed_utility
+                .checked_add(
+                    order.executed_utility(
+                        self.get_token_price(order.buy_token)
+                            .ok_or(SnappObjectiveError::TokenNotFound)?,
+                        *self
+                            .executed_buy_amounts
+                            .get(i)
+                            .ok_or(SnappObjectiveError::OrderBookMismatch)?,
+                        *self
+                            .executed_sell_amounts
+                            .get(i)
+                            .ok_or(SnappObjectiveError::OrderBookMismatch)?,
+                    )?,
+                )
+                .ok_or(SnappObjectiveError::TotalUtilityOverflow)?;
+        }
+
+        Ok(total_executed_utility)
+    }
+}
+
+/// Snapp specific extensions for `Order`. This extension trait provides the
+/// required functions for calculating the objective value used in the Snapp
+/// protocol which is needed for solution submission.
+pub trait SnappOrder {
+    /// Returns the executed utility of an order based on executed amounts
+    /// and closing prices.
+    ///
+    /// Note that utility is calculated with the following equasion:
+    /// `((exec_buy_amt * sell_amt - exec_sell_amt * buy_amt) * buy_price) /
+    /// sell_amt`
+    ///
+    /// # Returns
+    ///
+    /// `Ok(U256)` when the order executed utility is correctly calculated
+    /// and `Err` otherwise.
+    fn executed_utility(
+        &self,
+        buy_price: u128,
+        exec_buy_amt: u128,
+        exec_sell_amt: u128,
+    ) -> Result<U256, SnappObjectiveError>;
+}
+
+impl SnappOrder for Order {
+    fn executed_utility(
+        &self,
+        buy_price: u128,
+        exec_buy_amt: u128,
+        exec_sell_amt: u128,
+    ) -> Result<U256, SnappObjectiveError> {
+        let buy_price = u128_to_u256(buy_price);
+        let exec_buy_amt = u128_to_u256(exec_buy_amt);
+        let exec_sell_amt = u128_to_u256(exec_sell_amt);
+        let buy_amt = u128_to_u256(self.buy_amount);
+        let sell_amt = u128_to_u256(self.sell_amount);
+
+        // the utility is caculated in two parts, this is done to avoid overflows
+        let utility_with_error = exec_buy_amt
+            .checked_sub(
+                (exec_sell_amt * buy_amt)
+                    .checked_div(sell_amt)
+                    .ok_or(SnappObjectiveError::InvalidOrder)?,
+            )
+            .ok_or(SnappObjectiveError::LimitPriceNotRespected)?
+            * buy_price;
+        let utility_error = (((exec_sell_amt * buy_amt) % sell_amt) * buy_price) / sell_amt;
```

Comment:
Any reason we are not doing checked math here?

Reply:
It can't overflow. `exec_sell_amt * buy_amt` is at most `u128::MAX * u128::MAX` which always fits into a `u256`. `x % sell_amt` is at most `u128::MAX - 1` when `sell_amt` is `u128::MAX` and so `(x % sell_amt) * buy_price` is at most `(u128::MAX - 1) * u128::MAX` which also always fits into a `u256`. Dividing by `sell_amt` has already been checked as well so it is guaranteed to not be `0`.